File: src/ai/sentimentAnalyzer.ts
import { SentimentAspect, SentimentResult } from '../types';

const POSITIVE_WORDS = new Set(['amazing', 'great', 'excellent', 'delicious', 'tasty', 'fresh', 'friendly', 'fast', 'quick', 'clean', 'good', 'love', 'loved', 'perfect', 'affordable', 'generous', 'warm', 'polite', 'recommend', 'best', 'wonderful', 'juicy']);
const NEGATIVE_WORDS = new Set(['slow', 'bad', 'cold', 'rude', 'dirty', 'small', 'expensive', 'bland', 'late', 'worst', 'terrible', 'awful', 'overpriced', 'stale', 'disappointing', 'poor', 'undercooked', 'burnt', 'greasy']);

const ASPECT_KEYWORDS: Record<SentimentAspect['label'], Set<string>> = {
  Taste: new Set(['taste', 'tasty', 'flavor', 'flavour', 'delicious', 'bland', 'spicy', 'juicy']),
  Service: new Set(['service', 'staff', 'waiter', 'waitress', 'server', 'slow', 'rude', 'friendly', 'polite']),
  Portion: new Set(['portion', 'quantity', 'size', 'small', 'large', 'generous']),
  Price: new Set(['price', 'expensive', 'cheap', 'affordable', 'overpriced', 'value']),
  Cleanliness: new Set(['clean', 'dirty', 'hygiene', 'hygienic']),
  Ambience: new Set(['ambience', 'ambiance', 'atmosphere', 'decor', 'seating', 'music']),
};

function tokenize(text: string): string[] {
  return (text.toLowerCase().match(/[a-z']+/g) || []);
}

function sentenceSplit(text: string): string[] {
  return text.split(/[.!?]/).map((s) => s.trim()).filter(Boolean);
}

function sentimentOf(tokens: string[]): 'Positive' | 'Negative' | 'Neutral' {
  let pos = 0, neg = 0;
  tokens.forEach((t) => { if (POSITIVE_WORDS.has(t)) pos++; if (NEGATIVE_WORDS.has(t)) neg++; });
  if (pos > neg) return 'Positive';
  if (neg > pos) return 'Negative';
  return 'Neutral';
}

/** Lightweight aspect-based sentiment analysis (rule-based, transparent). */
export function analyzeSentiment(text: string): SentimentResult {
  const sentences = sentenceSplit(text).length ? sentenceSplit(text) : [text];
  const aspectSentiments: Record<string, ('Positive' | 'Negative' | 'Neutral')[]> = {};

  sentences.forEach((sentence) => {
    const tokens = tokenize(sentence);
    const s = sentimentOf(tokens);
    (Object.keys(ASPECT_KEYWORDS) as SentimentAspect['label'][]).forEach((aspect) => {
      if (tokens.some((t) => ASPECT_KEYWORDS[aspect].has(t))) {
        aspectSentiments[aspect] = aspectSentiments[aspect] || [];
        aspectSentiments[aspect].push(s);
      }
    });
  });

  const aspects: SentimentAspect[] = Object.entries(aspectSentiments).map(([label, sentiments]) => {
    const pos = sentiments.filter((s) => s === 'Positive').length;
    const neg = sentiments.filter((s) => s === 'Negative').length;
    const overall = pos > neg ? 'Positive' : neg > pos ? 'Negative' : 'Neutral';
    return { label: label as SentimentAspect['label'], sentiment: overall };
  });

  const allTokens = tokenize(text);
  const posCount = allTokens.filter((t) => POSITIVE_WORDS.has(t)).length;
  const negCount = allTokens.filter((t) => NEGATIVE_WORDS.has(t)).length;
  let overall: SentimentResult['overall'] = 'Neutral';
  if (posCount > 0 && negCount > 0) overall = 'Mixed';
  else if (posCount > negCount) overall = 'Positive';
  else if (negCount > posCount) overall = 'Negative';

  return { overall, aspects };
}

File: backend/app/ai/sentiment_analyzer.py
"""
Aspect-Based Sentiment Analysis (rule-based, transparent).

For the FYP prototype this uses a lightweight lexicon + aspect-keyword
approach rather than a heavyweight trained transformer, so it runs
instantly with no model download and is easy to explain in a defense.
The module is structured so it can be swapped for a Hugging Face
transformers pipeline later (see NOTE at the bottom) without touching
any caller code -- analyze_review() is the stable interface.
"""
import re
from typing import Dict, List

POSITIVE_WORDS = {
    "amazing", "great", "excellent", "delicious", "tasty", "fresh", "friendly",
    "fast", "quick", "clean", "good", "love", "loved", "perfect", "affordable",
    "generous", "warm", "polite", "recommend", "best", "wonderful", "juicy",
}
NEGATIVE_WORDS = {
    "slow", "bad", "cold", "rude", "dirty", "small", "expensive", "bland",
    "late", "worst", "terrible", "awful", "overpriced", "stale", "disappointing",
    "poor", "undercooked", "burnt", "greasy",
}

ASPECT_KEYWORDS = {
    "Taste": {"taste", "tasty", "flavor", "flavour", "delicious", "bland", "spicy", "juicy"},
    "Service": {"service", "staff", "waiter", "waitress", "server", "slow", "rude", "friendly", "polite"},
    "Portion": {"portion", "quantity", "size", "small", "large", "generous"},
    "Price": {"price", "expensive", "cheap", "affordable", "overpriced", "value"},
    "Cleanliness": {"clean", "dirty", "hygiene", "hygienic"},
    "Ambience": {"ambience", "ambiance", "atmosphere", "decor", "seating", "music"},
}


def _tokenize(text: str) -> List[str]:
    return re.findall(r"[a-zA-Z']+", text.lower())


def _sentence_split(text: str) -> List[str]:
    return [s.strip() for s in re.split(r"[.!?]", text) if s.strip()]


def _sentiment_of(tokens: List[str]) -> str:
    pos = sum(1 for t in tokens if t in POSITIVE_WORDS)
    neg = sum(1 for t in tokens if t in NEGATIVE_WORDS)
    if pos > neg:
        return "Positive"
    if neg > pos:
        return "Negative"
    return "Neutral"


def analyze_review(text: str) -> Dict:
    sentences = _sentence_split(text) or [text]
    aspect_results: Dict[str, List[str]] = {a: [] for a in ASPECT_KEYWORDS}

    for sentence in sentences:
        tokens = _tokenize(sentence)
        sentiment = _sentiment_of(tokens)
        for aspect, keywords in ASPECT_KEYWORDS.items():
            if any(t in keywords for t in tokens):
                aspect_results[aspect].append(sentiment)

    aspects = []
    for aspect, sentiments in aspect_results.items():
        if not sentiments:
            continue
        pos = sentiments.count("Positive")
        neg = sentiments.count("Negative")
        overall = "Positive" if pos > neg else "Negative" if neg > pos else "Neutral"
        aspects.append({"label": aspect, "sentiment": overall})

    all_tokens = _tokenize(text)
    overall_pos = sum(1 for t in all_tokens if t in POSITIVE_WORDS)
    overall_neg = sum(1 for t in all_tokens if t in NEGATIVE_WORDS)
    if overall_pos > 0 and overall_neg > 0:
        overall = "Mixed"
    elif overall_pos > overall_neg:
        overall = "Positive"
    elif overall_neg > overall_pos:
        overall = "Negative"
    else:
        overall = "Neutral"

    return {"overall": overall, "aspects": aspects}

# NOTE (future extension): to use a trained transformer instead, replace the
# body of analyze_review() with a call to a Hugging Face
# `pipeline("sentiment-analysis")` / aspect-based-sentiment model and keep
# the same return shape: {"overall": str, "aspects": [{"label","sentiment"}]}.
